### PR TITLE
add ability to specify custom DNS settings for a VM in the cluster

### LINKED ISSUE #2182: VM Nics should be able to specific custom DNS settings
/kind feature

**Describe the solution you'd like**
A user would be able to specify a custom DNS setting for a VM in the cluster.  This would be specified here:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/networkinterfaces/spec.go#L148

Setting is https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/virtualmachines?tabs=bicep#virtualmachinenetworkinterfaceconfigurationproperties

**Anything else you would like to add:**

This is needed for customers using gMSA with a custom domain.  The Machine it's self needs to have DNS configured to point at domain controller.

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`): 

/sig windows

### PR DESCRIPTION
**Pull request by nawazkh**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
- A user would be able to specify a custom DNS setting for a VM in the cluster.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2182 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
AzureMachineSpec has been updated with DNSServers field. Users get to specify custom DNS Settings for a VM in the cluster.
```

### CODE REVIEW

File: azure/services/networkinterfaces/spec.go
Comment (nawazkh) [inline azure/services/networkinterfaces/spec.go:133]: @jsturtevant Is it okay to be passing `<nil>` `dnsSettings` in case `s.DNSServers` list is empty?

Comment (jackfrancis) [reply azure/services/networkinterfaces/spec.go:133]: Are you concerned about a nil slice being evaluated as `!=  0`? In fact that's not the case in golang. See:

https://go.dev/play/p/E0ilKGlGiMg

File: azure/services/networkinterfaces/spec.go
Comment (nawazkh) [inline azure/services/networkinterfaces/spec.go:159]: Sorry to confuse, please ignore, I think I answered my own question as I was thinking over the returned `network.Interface` object.

I had meant to ask "if we are going to break anything later on if `&dnsSettings` is a `nil` struct in the below return type? 
Should we be appending an empty slice instead to `&dnsSettings.DNSServers` ?"

But I realized that `&dnsSettings` can be nil if `len(s.DNSServers) == 0`. 

I.e. below returned object is acceptable.
```
network.Interface{
		Location: to.StringPtr(s.Location),
		InterfacePropertiesFormat: &network.InterfacePropertiesFormat{
			EnableAcceleratedNetworking: s.AcceleratedNetworking,
			IPConfigurations:            &ipConfigurations,
			DNSSettings:                 &dnsSettings, // <- this gets evaluated to <nil> if len(s.DNSServers) == 0
			EnableIPForwarding:          to.BoolPtr(s.EnableIPForwarding),
		},
	}, nil
```

File: azure/services/networkinterfaces/spec.go
Comment (nawazkh) [reply azure/services/networkinterfaces/spec.go:133]: Sorry that I confused, I meant to ask https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2447/files#r932527885

Comment (jsturtevant) [reply azure/services/networkinterfaces/spec.go:133]: I think you got your answer :-)  this looks ok to me....

**Comment by nawazkh** [comment]
I see the same error as [above listed prow apidiff](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2447/pull-cluster-api-provider-azure-apidiff/1556987785515110400) on running `make apidiff` locally
```
sigs.k8s.io/cluster-api-provider-azure/azure/services/networkinterfaces
  Incompatible changes:
  - NICSpec: old is comparable, new is not
  Compatible changes:
  - NICSpec.DNSServers: added

sigs.k8s.io/cluster-api-provider-azure/azure/scope
  Incompatible changes:
  - (*MachineScope).InboundNatSpecs: changed from func(map[int32]struct{}) []sigs.k8s.io/cluster-api-provider-azure/azure.ResourceSpecGetter to func() []sigs.k8s.io/cluster-api-provider-azure/azure.ResourceSpecGetter
  .
.
.
```

Is there something else I should be taking care of since I am adding a new field(_DNSServers_) to the [NICSpec here](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2447/files#diff-9e4a0ba373e2b8988e791145728267650b5b2e8a395f3f02a24e43e839081b7aR48) ?

**Comment by jsturtevant** [comment]
I believe `pull-cluster-api-provider-azure-apidiff` isn't a required test and is expected to fail in this case.  It's a signal for the maintainers that this PR is changing the API.  I think you are all set for that test👌

**Review by mboersma — COMMENTED** [review]
This is looking good. I only had one general comment which is mostly a matter of taste and can be ignored.

Could you squash the commits together?

File: api/v1alpha3/azuremachine_conversion.go
Comment (nawazkh) [reply api/v1alpha3/azuremachine_conversion.go]: Addressed the suggestion via commit bd7030074cbd4d78912eedcff78b2970254a42be
Also, you made my day. The thought process behind _"Why does Go use signed integers to index slices/arrays when there are no negative indices?"_ is beautiful. Thanks for linking it here!

**Review by mboersma — APPROVED** [review]
/lgtm

**Comment by CecileRobertMichon** [comment]
@nawazkh can you please rebase and address the merge conflict when you get a chance?

**Comment by nawazkh** [comment]
Fixed merge conflicts and updated the PR @CecileRobertMichon @mboersma. Please take a look when you have a chance

**Comment by nawazkh** [comment]
> Fixed merge conflicts and updated the PR @CecileRobertMichon @mboersma. Please take a look when you have a chance

PR tests are failing because I havent updated the spec_test for the [new fields added in NICSpec](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2447/files#diff-9e4a0ba373e2b8988e791145728267650b5b2e8a395f3f02a24e43e839081b7aR51-R52)
Please hold off the review for now.

/hold

**Review by CecileRobertMichon — COMMENTED** [review]
@nawazkh please remove the hold if this is ready

`/hold cancel`

**Comment by CecileRobertMichon** [comment]
/lgtm
/approve

🎉

**Review by jackfrancis — APPROVED** [review]
lgtm

**Pull request merged**
